### PR TITLE
Add VSpin (spin spin) Wrapper

### PR DESCRIPTION
**Pull request by Ph1so**
All main centrifuge functionalities work:
- door operations
- locking/unlocking 
- switching buckets
- spin cycle
- rotate certain distance

still need:
- unit tests 
- edge case handling
- better annotations
- documentation

smaller todos:
- switch setup to use a list of payloads

### CODE REVIEW

File: pylabrobot/centrifuge/agilent_centrifuge.py [hunk 1]
```diff
@@ -0,0 +1,787 @@
...
+  async def unlock_bucket(self):
+    await self.send(b"\xaa\x02\x26\x00\x06\x2e")
+    await self.com()
+
+  async def go_to_bucket1(self):
+    new_position = (self.position + 4000).to_bytes(4, byteorder="little")
+    byte_string = b"\xaa\x01\xd4\x97" + new_position + b"\xc3\xf5\x28\x00\xd7\x1a\x00\x00"
+    sum_byte = (sum(byte_string)-0xaa)&0xff
+    byte_string += sum_byte.to_bytes(1, byteorder="little")
+    tx_payloads = [
```
Comment (jkhales) [inline pylabrobot/centrifuge/agilent_centrifuge.py:259]: omg, this is intense

**Review by jkhales — COMMENTED** [review]
really exciting PR 👀

**Comment by Ph1so** [comment]
> really exciting PR 👀 

I've been dreaming of this moment.

File: pylabrobot/centrifuge/agilent_centrifuge.py (hunk 1 above)
Comment (rickwierenga) [reply pylabrobot/centrifuge/agilent_centrifuge.py:259]: Yes Phi has been cooking (his first reverse engineering project)

**Comment by jkhales** [comment]
Also suggest renaming this PR: "Add AgilentCentrifuge wrapper"

Imperative verb helps understand, are we adding, removing, updating?

**Comment by Ph1so** [comment]
> Also suggest renaming this PR to something like: "Add AgilentCentrifuge wrapper"
> 
> 
> 
> Imperative verb helps understand, are we adding, removing, updating?

Good idea 👍

File: pylabrobot/centrifuge/agilent_centrifuge.py [hunk 2]
```diff
@@ -0,0 +1,787 @@
...
+      raise RuntimeError("Failed to write all bytes")
+    resp = await self.read_resp(timeout=read_timeout)
+    # s = ""
+    # c = ""
+    # for i, byte in enumerate(resp):
+    #   s += f" {byte:02x} " # Format byte as two-digit hexadecimal
+    # for i, byte in enumerate(cmd):
+    #   c += f" {byte:02x} " # Format byte as two-digit hexadecimal
+    # print(f"TX: {len(cmd)}-bytes -\t {c}")
+    # print(f"RX: {len(resp)}-bytes -\t {s}\n")
```
Comment (rickwierenga) [inline pylabrobot/centrifuge/agilent_centrifuge.py:88]: is this still needed?

File: pylabrobot/centrifuge/agilent_centrifuge.py [hunk 3]
```diff
@@ -0,0 +1,787 @@
...
+    # TODO: add functionality where if robot has been intialized before nothing needs to happen
+    for _ in range(3):
+      await self.configure_and_initialize()
+      await self.send(b"\xaa\x00\x21\x01\xff\x21")
+    await self.finish_setup()
+    await self.status_check()
+
+  async def stop(self):
+    await self.com()
+    await self.configure_and_initialize()
```
Comment (rickwierenga) [inline pylabrobot/centrifuge/agilent_centrifuge.py:39]: this is needed on stop?

File: pylabrobot/centrifuge/agilent_centrifuge.py [hunk 4]
```diff
@@ -0,0 +1,787 @@
...
+    resp = await self.send(b"\xaa\x01\x0e\x0f")
+    # s = []
+    # for byte in resp:
+    #   s.append(f"{byte:02x}")
+    # await self.com()
+    # self.position = int.from_bytes(bytes.fromhex("".join(s[1:4])), byteorder="little")
+    # self.homing_position = int.from_bytes(bytes.fromhex("".join(s[9:12])), byteorder="little")
+    # stat = f"{resp[0]:02x}"
+    # print(f"status: {stat} position: {self.position}  homing position: {self.homing_position}")
+    return resp
```
Comment (rickwierenga) [inline pylabrobot/centrifuge/agilent_centrifuge.py:228]: is this still needed?

File: pylabrobot/centrifuge/agilent_centrifuge.py [hunk 5]
```diff
@@ -0,0 +1,787 @@
...
+
+    self.current_bucket = 2
+
+  async def start_spin_cycle(
+  self,
+  g: Optional[float] = 500,
+  time_seconds: Optional[float] = 10,
+  acceleration: Optional[float] = 80,
+ ) -> None:
+    """Start a spin cycle."""
```
Comment (rickwierenga) [inline pylabrobot/centrifuge/agilent_centrifuge.py:624]: it would be nice to document these parameters

File: pylabrobot/centrifuge/agilent_centrifuge.py [hunk 6]
```diff
@@ -0,0 +1,787 @@
...
+    if acceleration < 0:
+      acceleration = acceleration*(-1)
```
Comment (rickwierenga) [inline pylabrobot/centrifuge/agilent_centrifuge.py:626]: does this mean opposite direction?

File: pylabrobot/centrifuge/agilent_centrifuge.py [hunk 7]
```diff
@@ -0,0 +1,787 @@
...
+    acceleration = min(acceleration, 100)
```
Comment (rickwierenga) [inline pylabrobot/centrifuge/agilent_centrifuge.py:627]: why not raise an error? that might be better than this 'hidden' fix

File: pylabrobot/centrifuge/agilent_centrifuge.py [hunk 8]
```diff
@@ -0,0 +1,787 @@
...
+
+    rpm = int((g/(1.118*(10**(-4))))**0.5)
+    base = int(107007 - 328*rpm + 1.13*(rpm**2))
+    rpm_b = (int(4481*rpm + 10852)).to_bytes(4, byteorder="little")
+    acc = (int(915*acceleration/100)).to_bytes(2, byteorder="little")
+    maxp = min((self.position + base + 4000*rpm//30*time_seconds), 4294967294)
+    position = maxp.to_bytes(4, byteorder="little")
```
Comment (rickwierenga) [inline pylabrobot/centrifuge/agilent_centrifuge.py:634]: wow

File: pylabrobot/centrifuge/agilent_centrifuge.py [hunk 9]
```diff
@@ -0,0 +1,787 @@
...
+# "aa 01 0e 0f",
+# "aa 01 0e 0f"
+#     ]
+
+    for tx in payloads:
+      if isinstance(tx, str):
+        byte_literal = bytes.fromhex(tx)
+        await self.send(byte_literal)
+      else:
+        await self.send(tx)
```
Comment (rickwierenga) [inline pylabrobot/centrifuge/agilent_centrifuge.py:769]: this code seems shared and might benefit from a (private) helper function

**Review by rickwierenga — COMMENTED** [review]
phenomenal work

initial rough review, i will also test code locally

File: pylabrobot/centrifuge/agilent_centrifuge.py (hunk 6 above)
Comment (Ph1so) [reply pylabrobot/centrifuge/agilent_centrifuge.py:626]: This is just to handle if a user puts a negative number for whatever reason. do you think its necessary?

File: pylabrobot/centrifuge/agilent_centrifuge.py (hunk 7 above)
Comment (Ph1so) [reply pylabrobot/centrifuge/agilent_centrifuge.py:627]: Yeah.. maybe raising an error is a better idea

File: pylabrobot/centrifuge/agilent_centrifuge.py (hunk 6 above)
Comment (rickwierenga) [reply pylabrobot/centrifuge/agilent_centrifuge.py:626]: i think it's better to have the user specify things explicitly where possible

File: pylabrobot/centrifuge/agilent_centrifuge.py (hunk 3 above)
Comment (Ph1so) [reply pylabrobot/centrifuge/agilent_centrifuge.py:39]: mhm. will need to figure out why

File: pylabrobot/centrifuge/agilent_centrifuge.py [hunk 10]
```diff
@@ -0,0 +1,399 @@
...
+    await self.send(b"\xaa\x01\x0e\x0f")
+
+  async def stop(self):
+    await self.send(b"\xaa\x02\x0e\x10")
+    await self.configure_and_initialize()
+    if self.dev:
+      self.dev.close()
+
+  async def get_status(self):
+    resp = await self.send(b"\xaa\x01\x0e\x0f")
```
Comment (jkhales) [inline pylabrobot/centrifuge/agilent_centrifuge.py:147]:
```suggestion
    resp = await self.send(b"\xaa\x01\x0e\x0f")
    if len(resp) == 0:
        raise IOError("Empty status from centrifuge")
```

seems like you assume this is non empty below when you set `self.status`

File: pylabrobot/centrifuge/agilent_centrifuge.py [hunk 11]
```diff
@@ -0,0 +1,399 @@
...
+        if end_byte_found or time.time() - start_time > timeout:
+          logger.warning("Timed out reading response")
+          break
+        await asyncio.sleep(0.0001)
+
+    logger.debug("Read %s", data.hex())
+    return data
+
+  async def send(self, cmd: Union[bytearray, bytes], read_timeout=0.2) -> bytes:
+    """Send a command to the centrifuge and return the response."""
```
Comment (jkhales) [inline pylabrobot/centrifuge/agilent_centrifuge.py:186]:
```suggestion
```
seems like a redundant comment

File: pylabrobot/centrifuge/agilent_centrifuge.py [hunk 12]
```diff
@@ -0,0 +1,399 @@
...
+
+  async def set_configuration_data(self):
+    """Set the device configuration data."""
+    self.dev.ftdi_fn.ftdi_set_latency_timer(16)
+    self.dev.ftdi_fn.ftdi_set_line_property(8, 1, 0)
+    self.dev.ftdi_fn.ftdi_setflowctrl(0)
+    self.dev.baudrate = 19200
+
+  async def initialize(self):
+    """Initialize the device."""
```
Comment (jkhales) [inline pylabrobot/centrifuge/agilent_centrifuge.py:221]:
```suggestion
```

File: pylabrobot/centrifuge/agilent_centrifuge.py [hunk 13]
```diff
@@ -0,0 +1,399 @@
...
+  async def send_payloads(self, payloads) -> None:
+    """Send a list of commands to the centrifuge."""
+    for tx in payloads:
+      if isinstance(tx, str):
+        byte_literal = bytes.fromhex(tx)
+        await self.send(byte_literal)
+      else:
+        await self.send(tx)
+
+# setup() and stop() helper functions: configure_and_initialize, set_configuration_data, initialize
```
Comment (jkhales) [inline pylabrobot/centrifuge/agilent_centrifuge.py:207]:
```suggestion
```

File: pylabrobot/centrifuge/agilent_centrifuge.py [hunk 14]
```diff
@@ -0,0 +1,399 @@
...
+  "aa 01 0b 0c",
+  "aa 01 e6 c8 00 b0 04 96 00 0f 00 4b 00 a0 0f 05 00 07",
+  byte_string
+  ]
+    await self.close_door()
+    await self.lock_door()
+
+    await self.send_payloads(move_bucket)
+
+    time.sleep(2)
```
Comment (jkhales) [inline pylabrobot/centrifuge/agilent_centrifuge.py:278]:
```suggestion
    await asyncio.sleep(2)
```
otherwise you will block the event loop

File: pylabrobot/centrifuge/agilent_centrifuge.py [hunk 15]
```diff
@@ -0,0 +1,399 @@
...
```
Comment (jkhales) [inline pylabrobot/centrifuge/agilent_centrifuge.py:310]:
```suggestion
    await asyncio.sleep(2)
```

File: pylabrobot/centrifuge/agilent_centrifuge.py [hunk 16]
```diff
@@ -0,0 +1,399 @@
...
+
+      >>> cf.start_spin_cycle(g = 1000, time_seconds = 300, acceleration = 100)
+    """
+
+    if acceleration < 1 or acceleration > 100:
+      raise CentrifugeOperationError("Acceleration must be within 1-100.")
+    if g < 1 or g > 1000:
+      raise CentrifugeOperationError("G-force must be within 1-1000")
+    if time_seconds < 1:
+      raise CentrifugeOperationError("Spin time must be atleast 1 second")
```
Comment (jkhales) [inline pylabrobot/centrifuge/agilent_centrifuge.py:344]:
```suggestion
      raise CentrifugeOperationError("Spin time must be at least 1 second")
```

File: pylabrobot/centrifuge/agilent_centrifuge.py [hunk 17]
```diff
@@ -0,0 +1,399 @@
...
+
+    await self.send_payloads(payloads)
+
+    await self.get_status()
+
+    while self.status == "08":
+      time.sleep(1)
+      await self.get_status()
+
+    time.sleep(2)
```
Comment (jkhales) [inline pylabrobot/centrifuge/agilent_centrifuge.py:383]:
```suggestion
      await asyncio.sleep(1)
      await self.get_status()

    await asyncio.sleep(2)
```

File: README.md [hunk 18]
```diff
@@ -77,6 +77,18 @@ lh.move_plate(lh.deck.get_resource("plate"), pr)
 data = await pr.read_luminescence()
 ```
 
+### Centrifuges
+
+PyLabRobot also supports centrifugation (spin spin spin), currently limited to the Agilent Centrifuge. This functionality allows users to automate centrifuging processes on any modern operating system. Here is a quick example of centrifugation at 800 g-force for 60 seconds:
```
Comment (rickwierenga) [inline README.md:82]: i know we make this mistake elsewhere to, but the readme would get very long if we always start with 'plr also supports x'. the paragraph above can be shortened to: "Spinning at 800 g for 60 seconds using an Agilent VSpin"

(i shall take care of other such instances)

File: pylabrobot/centrifuge/agilent_centrifuge.py [hunk 19]
```diff
@@ -0,0 +1,377 @@
...
+    await self.send(b"\xaa\x02\x0e\x10")
+    await self.configure_and_initialize()
+    if self.dev:
+      self.dev.close()
+
+  async def get_status(self):
+    resp = await self.send(b"\xaa\x01\x0e\x0f")
+    if len(resp) == 0:
+      raise IOError("Empty status from centrifuge")
+    return resp
```
Comment (jkhales) [inline pylabrobot/centrifuge/agilent_centrifuge.py:156]: Have you considered making an enum so the types of status's that can be returned is understandable via code?

Comment (Ph1so) [reply pylabrobot/centrifuge/agilent_centrifuge.py:156]: I'm a little confused. Are you suggesting having like a dictionary of different statuses?

Comment (jkhales) [reply pylabrobot/centrifuge/agilent_centrifuge.py:157]: That would work. I like enums, but @rickwierenga probably likes dicts.

Just right now, it's not obvious what can be return by the different calls to `get_status` and so what are we expecting to come back?

Related, also curious why we are calling it in various places when the return value isn't analyzed?

Comment (Ph1so) [reply pylabrobot/centrifuge/agilent_centrifuge.py:157]: get_status returns 14 bytes which contain information on the current position of the centrifuge, the homing position, and an "unknown" status (this status can represent idling, spinning, etc). Perhaps it would be good to document this in the function. 

Also, Rick just found that the random status checks were actually kinda pointless. so those are gone now.

Comment (jkhales) [reply pylabrobot/centrifuge/agilent_centrifuge.py:157]: yea, those would be really nice to document!
